### PR TITLE
Completely remove build_aten and use_aten

### PR DESCRIPTION
**Pull request by mingzhe09088**
Breaking out of #8338 to completely remove build_aten and use_aten.

### CODE REVIEW

**Review by orionr — APPROVED** [review]
One thing I'm debating is whether we want all of the `ANDROID` and `IOS` checks (`if (NOT ANDROID and NOT IOS)`, etc.), but I think we are likely to incrementally remove these checks over time, so it's okay. I'm guessing all of these are needed for iOS and Android builds to pass?

**Comment by mingzhe09088** [comment]
Yes. The checks for mobile is needed for Android and iOS builds to pass.

**Comment by bddppq** [comment]
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang3.8-rocm1.7.1-ubuntu16.04-build/10441/console
```
06:45:13 [ 91%] Linking HIP shared library ../lib/libcaffe2_hip.so
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAByteType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDACharType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDADoubleType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAFloatType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAGenerator.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAHalfType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAIntType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDALongType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAShortType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_RegisterCUDA.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDAByteType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDACharType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDADoubleType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDAFloatType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDAIntType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDALongType.h.o: File format not recognized
06:45:14 objdump: CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_SparseCUDAShortType.h.o: File format not recognized
07:18:20 CMakeFiles/caffe2_hip.dir/__/build_caffe2/aten/src/ATen/caffe2_hip_generated_CUDAByteType.h.o: file not recognized: File format not recognized
07:18:20 clang-7.0: error: linker command failed with exit code 1 (use -v to see invocation)
07:18:20 Died at /opt/rocm/hip/bin/hipcc line 565.
```

@iotamudelta @Jorghi12 Could you help taking a look?

**Comment by bddppq** [comment]
After investigation this is because `ATen_CUDA_SRCS` contains header files, and `hip_add_library` determines whether a file is header only based on property `HEADER_FILE_ONLY` and not look into the filename. Filtering out the headers from ATen_CUDA_SRCS has fixed the issue.

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -30,7 +30,6 @@ cmake --version
 pip install -r requirements.txt || true
 
 if [[ "$BUILD_ENVIRONMENT" == *rocm* ]]; then
-  export MAX_JOBS=4
```
Comment (bddppq) [inline .jenkins/pytorch/build.sh]: tested that this is not needed anymore after switching to bigcpu machines

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -539,19 +539,27 @@ if(BUILD_CAFFE2 OR BUILD_ATEN)
...
     list(APPEND Caffe2_HIP_DEPENDENCY_LIBS
       roc::rocblas)
+
+    # TODO: Currently pytorch hipify script uses a feature called
+    # "disabled_modules" that effectively ifdef out a file, but
+    # without doing extra processing in the callers, which results in
+    # some unresolved symbols in the shared lib
+    # (libcaffe2_hip.so). Remove this when all disabled_modules are
+    # eliminated.
+    set(CMAKE_EXE_LINKER_FLAGS "-Wl,--unresolved-symbols=ignore-in-shared-libs ${CMAKE_EXE_LINKER_FLAGS}")
```
Comment (bddppq) [inline cmake/Dependencies.cmake]: @Jorghi12 @iotamudelta Please remember to remove this once the unresolved symbols are eliminated.

**Comment by bddppq** [comment]
Alt Text

**Pull request closed without merging**